internal/processor/stream/processor.go

```diff
@@ -212,61 +217,89 @@ func (p *Processor) readBatch(
 // Callers must not access result concurrently with HandleStream.
 func (p *Processor) HandleStream(
 	ctx context.Context,
+	async bool,
 	baseEvent model.APMEvent,
 	reader io.Reader,
 	batchSize int,
 	processor model.BatchProcessor,
 	result *Result,
 ) error {
-	// Since processor.ProcessBatch can block for some time until all the batch
-	// is added to the modelindexer's cache and there isn't a fail-fast rejection,
-	// we want to cap how many in-flight requests are read at any time.
+	// Limits the number of concurrent batch decodes.
 	// Defaults to 200 (N), only allowing N requests to read and cache Y events
-	// (determined by batchSize) from the batch, effectively limitting the decoding
-	// concurrency to N batches at any time, which should be a good ceiling. The
-	// ceiling also reduces the contention on the modelindexer.activeMu.
-	select {
-	case p.sem <- struct{}{}:
-	case <-ctx.Done():
-		return ctx.Err()
+	// (determined by batchSize) from the batch.
+	// The ceiling may also reduce the contention on the modelindexer.activeMu.
+	// Clients can set a async to true which makes the processor process the
+	// events in the background. Returns with an error `publish.ErrFull` if the
+	// semaphore is full. When asynchronous processing is requested, the batches
+	// are decoded synchronously, but the batch is processed asynchronously.
+	if err := p.acquireLock(ctx, async); err != nil {
+		return err
 	}
-
 	sr := p.getStreamReader(reader)
-	defer func() {
-		sr.release()
-		<-p.sem
-	}()
 
 	// first item is the metadata object
 	if err := p.readMetadata(sr, &baseEvent); err != nil {
 		// no point in continuing if we couldn't read the metadata
+		sr.release()
+		p.releaseLock()
 		return err
 	}
 
+	defer func() {
+		sr.release()
+		if !async { // Release lock for sync processing.
+			p.releaseLock()
+		}
+	}()
+
 	sp, ctx := apm.StartSpan(ctx, "Stream", "Reporter")
 	defer sp.End()
 
+	var acquireLock bool
 	for {
+		// Async requests will re-acquire the lock after the first batch is
+		// scheduled to be processed. If the semaphore is full, then we return
+		// with `publish.ErrFull`.
+		if async {
+			if acquireLock {
+				if err := p.acquireLock(ctx, async); err != nil {
```

Comment:
I'm pretty sure there are some error cases where we're acquiring the lock, but never releasing it. e.g. if `readBatch` returns `io.EOF` but there are no events to decode, because it's the end of the stream.